<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from smartcar.urdf.xacro            | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="smart_robot_car">
  <!-- Base link -->
  <link name="base_link">
    <visual>
      <geometry>
        <box size="0.3 0.15 0.01"/>
        <!-- Base dimensions based on wheelbase width -->
      </geometry>
    </visual>
    <collision>
      <geometry>
        <box size="0.3 0.15 0.01"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="1.0"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.01"/>
    </inertial>
  </link>
  <!-- Chassis of the robot -->
  <link name="chassis_link">
    <visual>
      <geometry>
        <box size="0.3 0.2 0.05"/>
        <!-- Dimensions of the chassis -->
      </geometry>
    </visual>
    <collision>
      <geometry>
        <box size="0.3 0.2 0.05"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="5.0"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.1" ixy="0.0" ixz="0.0" iyy="0.1" iyz="0.0" izz="0.1"/>
    </inertial>
    <parent link="base_link"/>
    <material name="Red"/>
  </link>
  <!-- Fixed joint to connect base_link to chassis_link -->
  <joint name="base_to_chassis" type="fixed">
    <parent link="base_link"/>
    <child link="chassis_link"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
  </joint>
  <link name="front_left_wheel_link">
    <visual>
      <geometry>
        <cylinder length="0.025" radius="0.032"/>
        <!-- Wheel diameter: 0.064 / 2 -->
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.025" radius="0.032"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.2"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.01"/>
    </inertial>
    <material name="Black"/>
  </link>
  <joint name="front_left_wheel_joint" type="fixed">
    <parent link="front_left_wheel_intermediary_link"/>
    <child link="front_left_wheel_link"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <axis xyz="0 1 0"/>
  </joint>
  <link name="front_left_wheel_intermediary_link">
    <inertial>
      <mass value="0.2"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.01"/>
    </inertial>
  </link>
  <joint name="front_left_wheel_steer_joint" type="revolute">
    <parent link="base_link"/>
    <child link="front_left_wheel_intermediary_link"/>
    <origin rpy="1.57 0 0" xyz="0.1285 0.11 0"/>
    <axis xyz="0 0 1"/>
    <!-- Steering axis for front left wheel -->
    <limit effort="5.0" velocity="5.0"/>
  </joint>
  <link name="front_right_wheel_link">
    <visual>
      <geometry>
        <cylinder length="0.025" radius="0.032"/>
        <!-- Wheel diameter: 0.064 / 2 -->
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.025" radius="0.032"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.2"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.01"/>
    </inertial>
    <material name="Black"/>
  </link>
  <joint name="front_right_wheel_joint" type="fixed">
    <parent link="front_right_wheel_intermediary_link"/>
    <child link="front_right_wheel_link"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <axis xyz="0 1 0"/>
  </joint>
  <link name="front_right_wheel_intermediary_link">
    <inertial>
      <mass value="0.2"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.01"/>
    </inertial>
  </link>
  <joint name="front_right_wheel_steer_joint" type="revolute">
    <parent link="base_link"/>
    <child link="front_right_wheel_intermediary_link"/>
    <origin rpy="1.57 0 0" xyz="0.1285 -0.11 0"/>
    <axis xyz="0 0 1"/>
    <!-- Steering axis for front left wheel -->
    <limit effort="5.0" velocity="5.0"/>
  </joint>
  <link name="back_left_wheel_link">
    <visual>
      <geometry>
        <cylinder length="0.025" radius="0.032"/>
        <!-- Wheel diameter: 0.064 / 2 -->
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.025" radius="0.032"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.2"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.01"/>
    </inertial>
    <material name="Black"/>
  </link>
  <joint name="back_left_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="back_left_wheel_link"/>
    <origin rpy="1.57 0 0" xyz="-0.1285 0.11 0"/>
    <axis xyz="0 0 1"/>
    <limit effort="10.0" velocity="10.0"/>
  </joint>
  <link name="back_right_wheel_link">
    <visual>
      <geometry>
        <cylinder length="0.025" radius="0.032"/>
        <!-- Wheel diameter: 0.064 / 2 -->
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.025" radius="0.032"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.2"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.01"/>
    </inertial>
    <material name="Black"/>
  </link>
  <joint name="back_right_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="back_right_wheel_link"/>
    <origin rpy="1.57 0 0" xyz="-0.1285 -0.11 0"/>
    <axis xyz="0 0 1"/>
    <limit effort="10.0" velocity="10.0"/>
  </joint>
  <!-- LiDAR sensor -->
  <link name="lidar_link">
    <visual>
      <geometry>
        <cylinder length="0.1" radius="0.05"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.1" radius="0.05"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.2"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.01"/>
    </inertial>
    <material name="Blue"/>
  </link>
  <joint name="lidar_joint" type="fixed">
    <parent link="chassis_link"/>
    <child link="lidar_link"/>
    <origin rpy="0 0 0" xyz="0.15 0 0.1"/>
    <!-- Position of the LiDAR on top of the car -->
  </joint>
  <!-- IMU sensor link -->
  <link name="imu_link">
    <visual>
      <geometry>
        <box size="0.05 0.05 0.02"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <box size="0.05 0.05 0.02"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.01"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.0001" ixy="0" ixz="0" iyy="0.0001" iyz="0" izz="0.0001"/>
    </inertial>
  </link>
  <!-- IMU joint -->
  <joint name="imu_joint" type="fixed">
    <parent link="base_link"/>
    <child link="imu_link"/>
    <origin rpy="0 0 0" xyz="0 0 0.1"/>
  </joint>
  <!-- Gazebo IMU sensor plugin -->
  <gazebo>
    <sensor name="imu_sensor" type="imu">
      <always_on>1</always_on>
      <update_rate>50</update_rate>
      <imu>
        <topic>imu/data</topic>
        <gaussian_noise>0.01</gaussian_noise>
      </imu>
      <plugin filename="libgazebo_ros_imu.so" name="gazebo_ros_imu">
        <ros>
          <namespace>robot_namespace</namespace>
          <remapping>imu/data:=imu_data</remapping>
        </ros>
        <update_rate>50</update_rate>
        <robotNamespace>/robot_namespace</robotNamespace>
        <bodyName>imu_link</bodyName>
        <xyzOffset>0 0 0</xyzOffset>
        <rpyOffset>0 0 0</rpyOffset>
      </plugin>
    </sensor>
  </gazebo>
  <!-- Adding the Car Gazebo Plugin -->
  <gazebo>
    <plugin filename="libcar_gazebo_plugin.so" name="car_plugin">
    </plugin>
  </gazebo>
</robot>
